### PR TITLE
fix: electrs getting stuck broadcasting a tx already in chain

### PR DESCRIPTION
**Pull request by dpc**
https://github.com/fedimint/fedimint/actions/runs/4503632304/jobs/7927075007?pr=2004

Fix by considering this error as a success.

### CODE REVIEW

File: fedimint-bitcoind/src/bitcoincore_rpc.rs [hunk 1]
```diff
@@ -319,7 +320,24 @@ impl IBitcoindRpc for ElectrumClient {
...
+            match self.0.transaction_broadcast_raw(&bytes) {
+                Ok(_txid) => {}
+                Err(error) => {
+                    // debug
+                    debug!(?error, "electrs error when broadcasting transaction");
+                    if let electrum_client::Error::Protocol(serde_json::Value::Object(ref map)) =
+                        &error
+                    {
+                        if let Some(serde_json::Value::Number(ref n)) = map.get("code") {
+                            if n == &serde_json::Number::from(2) {
```
[the remark this answers is not in the document]
Comment (dpc) [reply fedimint-bitcoind/src/bitcoincore_rpc.rs:332]: I took it straight from the failing CI and assumed just like (couple dozens lines above this one) we can use the `code`. But now that I look into it https://electrumx-spesmilo.readthedocs.io/en/latest/protocol-methods.html I'm not so sure.

[the remark this answers is not in the document]
Comment (jkitman) [reply fedimint-bitcoind/src/bitcoincore_rpc.rs:332]: I see two errors

```
 fedimint-test-all> 00:09:50 2023-03-23T18:02:04.928146Z  INFO fedimint_bitcoind: Bitcoind error Electrum server error: {"code":2,"message":"Transaction already in block chain"}, retrying
fedimint-test-all> 00:09:51 [2023-03-23T18:02:05.523Z WARN  electrs::electrum] RPC blockchain.transaction.broadcast failed: failed to broadcast transaction: JSON-RPC error: RPC error response: RpcError { code: -27, message: "Transaction already in block chain", data: None }
```

[the remark this answers is not in the document]
Comment (dpc) [reply fedimint-bitcoind/src/bitcoincore_rpc.rs:332]: :confused: 

The part that I was looking at had mostly the `code: 2`. :/


I guess this needs https://github.com/fedimint/fedimint/issues/2017 to have any chance of really being fixed.

[the remark this answers is not in the document]
Comment (dpc) [reply fedimint-bitcoind/src/bitcoincore_rpc.rs:332]: That's what https://github.com/fedimint/fedimint/issues/2017 is more or less, yes.

[the remark this answers is not in the document]
Comment (jkitman) [reply fedimint-bitcoind/src/bitcoincore_rpc.rs:332]: This reproduces the failure https://github.com/fedimint/fedimint/pull/2018/files

[the remark this answers is not in the document]
Comment (dpc) [reply fedimint-bitcoind/src/bitcoincore_rpc.rs:332]: Awesome. Luckly, my code works. :D

```
2023-03-23T20:45:32.466852Z DEBUG fedimint_bitcoind::bitcoincore_rpc: Broadcasted transaction already in the blochchain. Ignoring error. error=Protocol(Object {"code": Number(2), "message": String
("Transaction already in block chain")})    
```

Though I'm not sure if `code = 2` is selective enough, or could it include some other errors too :shrug: .

But the esplora one is borked, returns just 400, and I don't see anything to help tell what went wrong.

**Comment by dpc** [comment]
Let's go with https://github.com/fedimint/fedimint/pull/2021 instead?

File: fedimint-bitcoind/src/bitcoincore_rpc.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (jkitman) [reply fedimint-bitcoind/src/bitcoincore_rpc.rs:332]: You could try to check if the transaction exists before broadcasting, but you still might hit a race condition with another peer who is broadcasting.

**Pull request closed without merging**